### PR TITLE
Improved monster movement and spawning is now in seperate class

### LINKED ISSUE #15: [Taking into account the User's location] Monster range and Player range
# Blocker*:
[Drawing the location of monsters near the User] Implement functionality to monster/s causing them to move closer to the user #9
Player base class #12
Monster base class #13 

## Description:
The purpose of  this issue is to implement the functionality to add "range" of monster/s or player ( circle range ).
Which after monster enter the player 'range', the player should lose one life.

## Objective:
- Look if there is already life system implemented, if not add it.
- Add player range ( circle ), which will be centered at user current location and will be placed in real-time.
- Add functionality that if monster enter said circle, the player will lose one life.

Update:

- Improve the 3rd type of monster that follows user. Implement their functionality.

## Additional Information:
Collaborate with the team in case of any questions or problems.
Test the functionality thoroughly to ensure there are no bugs related to this issue.

* blocker is a task that blocks the development of the current task, meaning this task can't be started if the blocker isn't finished.

### PR DESCRIPTION
**Pull request by MBigaj**
MapsActivity.kt:
	Removed spawning from  initial activity and moved to seperate
	class

Player now includes the text view for healthbar

### CODE REVIEW

File: app/src/main/java/com/example/fivenightsescape/MapsActivity.kt [hunk 1]
```diff
@@ -129,15 +109,23 @@ class MapsActivity : AppCompatActivity(), OnMapReadyCallback {
...
+            player = this.player,
+            mMap = mMap,
+            difficulty = intent.getStringExtra("selectedDifficulty")
+        )
 
         val handler = Handler(Looper.getMainLooper())
 
         handler.postDelayed(object : Runnable {
             override fun run() {
                 currentPlayerLocation?.let { location ->
```
Comment (PKrystian) [inline app/src/main/java/com/example/fivenightsescape/MapsActivity.kt]: Parameter 'location' is never used, could be renamed to _

File: app/src/main/java/com/example/fivenightsescape/monster/MonsterSpawner.kt [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        this.monsterController.monsterAlert()
+    }
+
+    private fun spawnMonster(monsterType: String)
+    {
+        val randomLatitude = randomPosition()
+        val randomLongitude = randomPosition()
+
+        when (monsterType) {
```
Comment (PKrystian) [inline app/src/main/java/com/example/fivenightsescape/monster/MonsterSpawner.kt:61]: Code repetition in spawnMonster – can we reduce redundancy for clarity?
Differences in types are MonsterStanding, Moving, Wandering in `this.monsterController.addMonster(xxxxxxx`, everything else is repeating.
Like something similar to addMonster method.
```kt
when (this.difficulty) {
  DIFFICULTY_EASY -> spawnMonsterForDifficulty(dice, EASY_SPAWNER_BASIC, EASY_SPAWNER_MODIFIER)
  DIFFICULTY_MEDIUM -> spawnMonsterForDifficulty(dice, MEDIUM_SPAWNER_BASIC, MEDIUM_SPAWNER_MODIFIER)
  DIFFICULTY_HARD -> spawnMonsterForDifficulty(dice, HARD_SPAWNER_BASIC, HARD_SPAWNER_MODIFIER)
  else -> {
    // do some log error
  }
}
```

File: app/src/main/java/com/example/fivenightsescape/player/Player.kt [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
+class Player(var position: LatLng, val healthBar: TextView): ViewModel() {
     private var health: Int = PLAYER_HEALTH
+
     var location: Location = Location("")
 
     init {
         this.location.latitude = this.position.latitude
         this.location.longitude = this.position.longitude
+
+        this.healthBar.text = this.health.toString()
```
Comment (PKrystian) [inline app/src/main/java/com/example/fivenightsescape/player/Player.kt]: Should be: (for both / one below)
```kt
"Health: " + this.health.toString()
```
as user won't know what numbers stand for.
Screenshot 2024-01-18 001035

File: app/src/main/java/com/example/fivenightsescape/player/Player.kt [hunk 4]
```diff
@@ -1,18 +1,22 @@
...
 import android.location.Location
+import android.widget.TextView
 import androidx.lifecycle.ViewModel
 import com.google.android.gms.maps.model.LatLng
 
-const val PLAYER_HEALTH = 3
+private const val PLAYER_HEALTH = 3
 
-class Player(var position: LatLng): ViewModel() {
+class Player(var position: LatLng, val healthBar: TextView): ViewModel() {
```
Comment (PKrystian) [inline app/src/main/java/com/example/fivenightsescape/player/Player.kt]: This field leaks a context object.
You can fix it by changing it to:
```kt
class Player(var position: LatLng, healthBar: TextView) : ViewModel() {
  private var health: Int = PLAYER_HEALTH // This line stays the same
  private val healthBarRef: WeakReference<TextView> = WeakReference(healthBar)
  private val healthBar: TextView?
  get() = healthBarRef.get()
```
add `?` to both `this.healthBar.text = this.health.toString()`:
```kt
this.healthBar?.text = this.health.toString()
```

**Review by PKrystian — CHANGES_REQUESTED** [review]
Found 2 big issues:

- Having less than 0 health doesn't end the game ( return to menu ).
-  One 'attack' from monster changes player health to '-1':
Screenshot 2024-01-18 001147

If pull request didn't change/add these functionalities, create an issue/bugfix about them, if not try to fix it.

And other small issues in review comments.

File: app/src/main/java/com/example/fivenightsescape/monster/MonsterSpawner.kt (hunk 2 above)
Comment (MBigaj) [reply app/src/main/java/com/example/fivenightsescape/monster/MonsterSpawner.kt:61]: Unfortunately it's impossible with how it is constructed now, we have to initialize monsters with their respective types, cannot create constructor builder

**Review by PKrystian — APPROVED** [review]
GJ!

**Pull request merged**